src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,190 @@
+            {
+                var downloadRootPath = Path.Combine(buildArtifact.Resource.Data, buildArtifact.Name);
+                await this.CopyFileShareAsync(downloadRootPath, Path.Combine(downloadParameters.TargetDirectory, buildArtifact.Name), downloadParameters.MinimatchFilters, cancellationToken);
+            }
+        }
+
+        public async Task PublishArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken)
+        {
+            await PublishArtifactUsingRobocopyAsync(this.context, new HostContext(this.hostType), sourcePath, destPath, cancellationToken);
+        }
+
+        private async Task CopyFileShareAsync(string downloadRootPath, string destPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken)
+        {
+            minimatchPatterns = minimatchPatterns.Select(pattern => Path.Combine(downloadRootPath, pattern));
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, this.tracer);
+            await DownloadFileShareArtifactAsync(downloadRootPath, destPath, defaultParallelCount, cancellationToken, minimatcherFuncs);
+        }
+        
+        private async Task PublishArtifactUsingRobocopyAsync(AgentTaskPluginExecutionContext executionContext, IHostContext hostContext, string dropLocation, string downloadFolderPath, CancellationToken cancellationToken)
+        {
+            executionContext.Output(StringUtil.Loc("PublishingArtifactUsingRobocopy"));
+            using (var processInvoker = hostContext.CreateService<IProcessInvoker>())
+            {
+                // Save STDOUT from worker, worker will use STDOUT report unhandle exception.
+                processInvoker.OutputDataReceived += delegate (object sender, ProcessDataReceivedEventArgs stdout)
+                {
+                    if (!string.IsNullOrEmpty(stdout.Data))
+                    {
+                        executionContext.Output(stdout.Data);
+                    }
+                };
+
+                // Save STDERR from worker, worker will use STDERR on crash.
+                processInvoker.ErrorDataReceived += delegate (object sender, ProcessDataReceivedEventArgs stderr)
+                {
+                    if (!string.IsNullOrEmpty(stderr.Data))
+                    {
+                        executionContext.Error(stderr.Data);
+                    }
+                };
+
+                var trimChars = new[] { '\\', '/' };
+
+                dropLocation = Path.Combine(dropLocation.TrimEnd(trimChars));
+                downloadFolderPath = downloadFolderPath.TrimEnd(trimChars);
+
+                string robocopyArguments = "\"" + dropLocation + "\" \"" + downloadFolderPath + "\" * /E /COPY:DA /NP /R:3";
+
+                int exitCode = await processInvoker.ExecuteAsync(
+                        workingDirectory: "",
+                        fileName: "robocopy",
```

Comment:
>robocopy [](start = 35, length = 8)

consider using [https://github.com/tjscience/RoboSharp](https://github.com/tjscience/RoboSharp) in the future